openedx/features/student_account/tests/test_set_opt_in_and_affiliate_user_organization.py

```diff
@@ -0,0 +1,135 @@
+        """
+        Test that when no organization related data is present in the request, the
+        form cleans the data without raising any validation errors, passes the
+        data to set_opt_in_and_affiliate_user_organization method, does not
+        fetch an existing organization or create a new organization,
+        and then saves the UserExtendedProfile and EmailPreferences
+        """
+
+        request = RequestFactory().post('/user_api/v1/account/registration/',
+                                        self.request_data)
+        params = dict(request.POST.copy().items())
+        params['name'] = '{} {}'.format(params.get('first_name'), params.get('last_name'))
+
+        form = AccountCreationFormCustom(data=params,
+                                         extended_profile_fields=None,
+                                         do_third_party_auth=False)
+        form.is_valid()
+        set_opt_in_and_affiliate_user_organization(self.user, form)
+
+        org_get_or_create_method.assert_not_called()
+        user_extended_profile_create_method.assert_called_once_with(user=self.user, **{})
+        email_preferences_create_method.assert_called_once_with(user=self.user, opt_in=form.cleaned_data.get('opt_in'))
+
+    @patch('openedx.features.student_account.helpers.EmailPreference.objects.create')
+    @patch('openedx.features.student_account.helpers.UserExtendedProfile.objects.create')
+    def test_user_with_new_organization(self, user_extended_profile_create_method, email_preferences_create_method):
+        """
+        Test that when the user gives a new organization fields in the request,
+        the form does not raise any validation error.
+
+        A new organization is created, and the user is made first_learner for that
+        organization in the UserExtendedProfile and the EmailPreferences are
+        also set for that user
+        """
+
+        self.request_data['org_name'] = 'new_organization'
+        self.request_data['org_size'] = '5-10'
+        self.request_data['org_type'] = 'IWRNS'
+
+        request = RequestFactory().post('/user_api/v1/account/registration/', self.request_data)
+        params = dict(request.POST.copy().items())
+        params['name'] = '{} {}'.format(params.get('first_name'), params.get('last_name'))
+
+        form = AccountCreationFormCustom(data=params,
+                                         extended_profile_fields=None,
+                                         do_third_party_auth=False)
+        form.is_valid()
+
+        set_opt_in_and_affiliate_user_organization(self.user, form)
+        created_organization = Organization.objects.filter(label=form.cleaned_data.get('org_name')).first()
+
+        user_extended_profile_data = {
+            'is_first_learner': True,
+            'organization_id': created_organization.id
+        }
+
+        user_extended_profile_create_method.assert_called_once_with(user=self.user, **user_extended_profile_data)
```

Comment:
Lets add assertion that `created_organization` has same `org_size`, `org_type` and `org_type` what we have in `request_data`